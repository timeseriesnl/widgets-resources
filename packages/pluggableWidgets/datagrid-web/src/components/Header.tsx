import { createElement, Dispatch, ReactElement, SetStateAction, DragEvent, DragEventHandler, useCallback } from "react";
import { ColumnInstance, HeaderGroup, IdType, SortingRule } from "react-table";
import classNames from "classnames";
import { FontAwesomeIcon } from "@fortawesome/react-fontawesome";
import { faLongArrowAltDown, faLongArrowAltUp, faArrowsAltV } from "@fortawesome/free-solid-svg-icons";
import { ColumnResizer } from "./ColumnResizer";

export interface HeaderProps<D extends object> {
    className?: string;
    column: HeaderGroup<D>;
    sortable: boolean;
    resizable: boolean;
    filterable: boolean;
    draggable: boolean;
    dragOver: string;
    hidable: boolean;
    isDragging?: boolean;
    preview?: boolean;
    setColumnOrder: (updater: Array<IdType<D>>) => void;
    setColumnWidth: (width: number) => void;
    setDragOver: Dispatch<SetStateAction<string>>;
    setIsDragging: Dispatch<SetStateAction<boolean>>;
    setSortBy: Dispatch<SetStateAction<Array<SortingRule<object>>>>;
    sortBy: Array<SortingRule<object>>;
    visibleColumns: Array<ColumnInstance<D>>;
}

export function Header<D extends object>(props: HeaderProps<D>): ReactElement {
    const canSort = props.sortable && props.column.canSort;
    const canDrag = props.draggable && (props.column.canDrag ?? false);
    const draggableProps = useDraggable(
        canDrag,
        props.visibleColumns,
        props.setColumnOrder,
        props.setDragOver,
        props.setIsDragging
    );

    const { style, ...rest } = props.column.getHeaderProps();

    const [sortProperties] = props.sortBy;
    const isSorted = sortProperties && sortProperties.id === props.column.id;
    const isSortedDesc = isSorted && sortProperties.desc;

    const sortIcon = canSort
        ? isSorted
            ? isSortedDesc
                ? faLongArrowAltDown
                : faLongArrowAltUp
            : faArrowsAltV
        : undefined;

    const caption = props.column.render("Header") as string;

    const onSortBy = (): void => {
        /**
         * Always analyse previous values to predict the next
         * 1 - !isSorted turns to asc
         * 2 - isSortedDesc === false && isSorted turns to desc
         * 3 - isSortedDesc === true && isSorted turns to unsorted
         * If multisort is allowed in the future this should be changed to append instead of just return a new array
         */
        if (!isSorted) {
            props.setSortBy([{ id: props.column.id, desc: false }]);
        } else if (isSorted && !isSortedDesc) {
            props.setSortBy([{ id: props.column.id, desc: true }]);
        } else {
            props.setSortBy([]);
        }
        // TODO: Find a way to do sorting better
    };

    return (
        <div
            className={classNames("th", {
                "hidden-column-preview": props.preview && props.hidable && props.column.hidden
            })}
            {...rest}
            style={{
                ...style,
                ...(!props.sortable || !props.column.canSort ? { cursor: "unset" } : undefined)
            }}
            title={caption}
        >
            <div
                id={props.column.id}
                className={classNames(
                    "column-container",
                    canDrag && props.column.id === props.dragOver ? "dragging" : ""
                )}
                {...draggableProps}
            >
                <div
                    id={props.column.id}
                    className={classNames("column-header", canSort ? "clickable" : "", props.className)}
                    style={{ pointerEvents: props.isDragging ? "none" : undefined }}
                    onClick={canSort ? onSortBy : undefined}
                    onKeyDown={
                        canSort
                            ? e => {
                                  if (e.key === "Enter" || e.key === " ") {
                                      e.preventDefault();
                                      onSortBy();
                                  }
                              }
                            : undefined
                    }
                    role={canSort ? "button" : undefined}
                    tabIndex={canSort ? 0 : undefined}
                >
                    <span>{caption.length > 0 ? caption : "\u00a0"}</span>
                    {sortIcon && <FontAwesomeIcon icon={sortIcon} />}
                </div>
                {props.filterable && props.column.customFilter ? props.column.customFilter : null}
            </div>
            {props.resizable && props.column.canResize && <ColumnResizer setColumnWidth={props.setColumnWidth} />}
        </div>
    );
}

function useDraggable<D extends object>(
    columnsDraggable: boolean,
    visibleColumns: Array<ColumnInstance<D>>,
    setColumnOrder: (updater: ((columnOrder: Array<IdType<D>>) => Array<IdType<D>>) | Array<IdType<D>>) => void,
    setDragOver: Dispatch<SetStateAction<string>>,
    setIsDragging: Dispatch<SetStateAction<boolean>>
): {
    draggable?: boolean;
    onDragStart?: DragEventHandler;
    onDragOver?: DragEventHandler;
    onDrop?: DragEventHandler;
    onDragEnter?: DragEventHandler;
    onDragEnd?: DragEventHandler;
} {
    const handleDragStart = useCallback((e: DragEvent<HTMLDivElement>): void => {
        setIsDragging(true);
        const { id } = e.target as HTMLDivElement;
        e.dataTransfer.setData("colDestination", id);
    }, []);

    const handleDragOver = useCallback((e: DragEvent<HTMLDivElement>): void => {
        e.preventDefault();
    }, []);

    const handleDragEnter = useCallback((e: DragEvent<HTMLDivElement>): void => {
        const { id } = e.target as HTMLDivElement;
        const colDestination = e.dataTransfer.getData("colDestination");
        if (id !== colDestination) {
            setDragOver(id);
        }
    }, []);

    const handleDragEnd = useCallback((): void => {
        setIsDragging(false);
        setDragOver("");
    }, []);

    const handleOnDrop = useCallback(
        (e: DragEvent<HTMLDivElement>): void => {
            handleDragEnd();
            const { id: colOrigin } = e.target as HTMLDivElement;
            const colDestination = e.dataTransfer.getData("colDestination");

            const toIndex = visibleColumns.findIndex((col: ColumnInstance<D>) => col.id === colOrigin);
            const fromIndex = visibleColumns.findIndex((col: ColumnInstance<D>) => col.id === colDestination);

            if (toIndex !== fromIndex) {
                const newOrder = [...visibleColumns.map(column => column.id)];
                newOrder.splice(fromIndex, 1);
                newOrder.splice(toIndex, 0, colDestination);
                setColumnOrder(newOrder);
            }
        },
        [visibleColumns]
    );

    return columnsDraggable
        ? {
              draggable: true,
              onDragStart: handleDragStart,
              onDragOver: handleDragOver,
              onDrop: handleOnDrop,
              onDragEnter: handleDragEnter,
              onDragEnd: handleDragEnd
          }
        : {};
}
