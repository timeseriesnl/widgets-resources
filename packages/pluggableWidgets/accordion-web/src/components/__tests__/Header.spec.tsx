import { createElement, PropsWithChildren } from "react";
import { shallow } from "enzyme";

import { Header, HeaderProps } from "../Header";

describe("Header", () => {
    let defaultHeaderProps: PropsWithChildren<HeaderProps>;

    beforeEach(() => {
        defaultHeaderProps = {
            heading: "headingThree",
            children: "Text"
        };
    });

    it("renders h1", () => {
        const headerWrapper = shallow(<Header {...defaultHeaderProps} heading="headingOne" />);

        expect(headerWrapper).toMatchSnapshot();
    });

    it("renders h2", () => {
        const headerWrapper = shallow(<Header {...defaultHeaderProps} heading="headingTwo" />);

        expect(headerWrapper).toMatchSnapshot();
    });

    it("renders h3", () => {
        const headerWrapper = shallow(<Header {...defaultHeaderProps} />);

        expect(headerWrapper).toMatchSnapshot();
    });

    it("renders h4", () => {
        const headerWrapper = shallow(<Header {...defaultHeaderProps} heading="headingFour" />);

        expect(headerWrapper).toMatchSnapshot();
    });

    it("renders h5", () => {
        const headerWrapper = shallow(<Header {...defaultHeaderProps} heading="headingFive" />);

        expect(headerWrapper).toMatchSnapshot();
    });

    it("renders h6", () => {
        const headerWrapper = shallow(<Header {...defaultHeaderProps} heading="headingSix" />);

        expect(headerWrapper).toMatchSnapshot();
    });
});
