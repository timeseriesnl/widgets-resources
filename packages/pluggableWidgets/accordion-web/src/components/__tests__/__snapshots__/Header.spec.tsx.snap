// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Header renders h1 1`] = `
<h1>
  Text
</h1>
`;

exports[`Header renders h2 1`] = `
<h2>
  Text
</h2>
`;

exports[`Header renders h3 1`] = `
<h3>
  Text
</h3>
`;

exports[`Header renders h4 1`] = `
<h4>
  Text
</h4>
`;

exports[`Header renders h5 1`] = `
<h5>
  Text
</h5>
`;

exports[`Header renders h6 1`] = `
<h6>
  Text
</h6>
`;
